index.d.ts

```diff
@@ -0,0 +1,269 @@
+// Type definitions for JSZip 3.1
+// Project: http://stuk.github.com/jszip/, https://github.com/stuk/jszip
+// Definitions by: mzeiher <https://github.com/mzeiher>, forabi <https://github.com/forabi>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
+
+/// <reference types="node" />
+
+interface JSZipSupport {
+    arraybuffer: boolean;
+    uint8array: boolean;
+    blob: boolean;
+    nodebuffer: boolean;
+}
+
+type Compression = 'STORE' | 'DEFLATE';
+
+interface Metadata  {
+    percent: number;
+    currentFile: string;
+}
+
+type OnUpdateCallback = (metadata: Metadata) => void;
+
+interface InputByType {
+    base64: string;
+    string: string;
+    text: string;
+    binarystring: string;
+    array: number[];
+    uint8array: Uint8Array;
+    arraybuffer: ArrayBuffer;
+    blob: Blob;
+    stream: NodeJS.ReadableStream;
+}
+
+interface OutputByType {
+    base64: string;
+    text: string;
+    binarystring: string;
+    array: number[];
+    uint8array: Uint8Array;
+    arraybuffer: ArrayBuffer;
+    blob: Blob;
+    nodebuffer: Buffer;
+}
+
+
+interface CompressedObject {
+    compressedSize: number;
+    uncompressedSize: number;
+    crc32: number;
+    compression: object;
+    compressedContent: string|ArrayBuffer|Uint8Array|Buffer;
+}
+
+type InputFileFormat = InputByType[keyof InputByType];
+
+declare namespace JSZip {
+    type InputType = keyof InputByType;
+
+    type OutputType = keyof OutputByType;
+
+    interface JSZipObject {
+        name: string;
+        dir: boolean;
+        date: Date;
+        comment: string;
+        /** The UNIX permissions of the file, if any. */
+        unixPermissions: number | string | null;
+        /** The UNIX permissions of the file, if any. */
+        dosPermissions: number | null;
+        options: JSZipObjectOptions;
+
+        _data: CompressedObject;
```

Comment:
This is private, so I don't think it should be part of the TS interface.

Reply:
Yes, but sometimes we need to get the compression size and origin file's size.

Reply:
Hmm. I'm not comfortable explicitly exposing deliberately private properties.

Could you remove this property from this PR, and open another PR that exposes this property explicitly?